tools/kani-cov/src/coverage.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module defines coverage-oriented data structures shared among
+//! subcommands and other utilities like the Rust tree-sitter.
+
+use console::style;
+use serde_derive::{Deserialize, Serialize};
+use std::path::{Path, PathBuf};
+use std::{collections::BTreeMap, fmt::Display};
+use std::{fmt, fs};
+use tree_sitter::{Node, Parser};
+
+pub type LineResults = Vec<(usize, Option<(u32, MarkerInfo)>)>;
+
+/// The possible outcomes for a Kani check.
+///
+/// Note: This data structure should not be duplicated in Kani -
+/// <https://github.com/model-checking/kani/issues/3541>
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "UPPERCASE")]
+pub enum CheckStatus {
+    Failure,
+    Covered,   // for `code_coverage` properties only
+    Satisfied, // for `cover` properties only
+    Success,
+    Undetermined,
+    Unreachable,
+    Uncovered,     // for `code_coverage` properties only
+    Unsatisfiable, // for `cover` properties only
+}
+
+/// Kani coverage check.
+///
+/// Note: This data structure should not be duplicated in Kani -
+/// <https://github.com/model-checking/kani/issues/3541>
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CoverageCheck {
+    pub function: String,
+    term: CoverageTerm,
+    pub region: CoverageRegion,
+    pub status: CheckStatus,
+}
+
+// Note: This `impl` should not be duplicated in Kani -
+// <https://github.com/model-checking/kani/issues/3541>
+impl std::fmt::Display for CheckStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let check_str = match self {
+            CheckStatus::Satisfied => style("SATISFIED").green(),
+            CheckStatus::Success => style("SUCCESS").green(),
+            CheckStatus::Covered => style("COVERED").green(),
+            CheckStatus::Uncovered => style("UNCOVERED").red(),
+            CheckStatus::Failure => style("FAILURE").red(),
+            CheckStatus::Unreachable => style("UNREACHABLE").yellow(),
+            CheckStatus::Undetermined => style("UNDETERMINED").yellow(),
+            CheckStatus::Unsatisfiable => style("UNSATISFIABLE").yellow(),
+        };
+        write!(f, "{check_str}")
+    }
+}
+
+/// Raw Kani coverage results.
+///
+/// Note: This data structure should not be duplicated in Kani -
+/// <https://github.com/model-checking/kani/issues/3541>
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CoverageResults {
+    pub data: BTreeMap<String, Vec<CoverageCheck>>,
+}
+
+/// Aggregated coverage results.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CombinedCoverageResults {
+    pub data: BTreeMap<String, Vec<(String, Vec<CovResult>)>>,
+}
```

Comment:
Wouldn't be better to use a `HashMap` here? Does the order matter? I also imagine this data could be quite large.

Reply:
The order doesn't really matter. Changed to `HashMap`.